File: leetcode150/TwoPointer/125_Valid_Palindrome.java
/**
 * Problem: 125. Valid Palindrome
 * A phrase is a palindrome if, after converting all uppercase letters into lowercase letters
 * and removing all non-alphanumeric characters, it reads the same forward and backward.
 * Alphanumeric characters include letters and numbers.
 *
 * Example 1:
 * Input: s = "A man, a plan, a canal: Panama"
 * Output: true
 * Explanation: "amanaplanacanalpanama" is a palindrome.
 *
 * Example 2:
 * Input: s = "race a car"
 * Output: false
 * Explanation: "raceacar" is not a palindrome.
 *
 * Example 3:
 * Input: s = " "
 * Output: true
 * Explanation: s is an empty string "" after removing non-alphanumeric characters.
 * Since an empty string reads the same forward and backward, it is a palindrome.
 */

/*
Approach M-1:
1. Replace all non-alphanumeric characters using `replaceAll` with regex "[^a-zA-Z0-9]".
2. Convert the string to lowercase using `toLowerCase`.
3. Use two pointers (`left` and `right`) to compare characters from both ends of the string.
4. Increment `left` and decrement `right` if characters match; otherwise, return false.
5. If pointers meet or cross without mismatches, return true.

Time Complexity: O(n) - One pass for replaceAll, one pass for toLowerCase, and one for the while loop.
Space Complexity: O(n) - Intermediate strings created by replaceAll and toLowerCase.
*/

// class Solution {
//     public boolean isPalindrome(String s) {
//         // Remove all non-alphanumeric characters
//         s = s.replaceAll("[^a-zA-Z0-9]", "");

//         // Convert to lowercase
//         s = s.toLowerCase();

//         // Initialize two pointers
//         int left = 0;
//         int right = s.length() - 1;

//         // Compare characters from both ends
//         while (left < right) {
//             if (s.charAt(left) == s.charAt(right)) {
//                 left++;
//                 right--;
//             } else {
//                 return false;
//             }
//         }

//         // If no mismatches, return true
//         return true;
//     }
// }

/*
Approach M-2:
1. Use two pointers (`left` and `right`) to traverse the string.
2. Skip non-alphanumeric characters using `Character.isLetterOrDigit`.
3. Compare characters at the `left` and `right` pointers in a case-insensitive manner using `Character.toLowerCase`.
4. Increment `left` and decrement `right` if characters match; otherwise, return false.
5. If pointers meet or cross without mismatches, return true.

Time Complexity: O(n) - Single pass of the string with two pointers.
Space Complexity: O(1) - No additional memory is used beyond a few variables.
*/

class Solution {
    public boolean isPalindrome(String s) {
        // Initialize two pointers
        int left = 0;
        int right = s.length() - 1;

        while (left < right) {
            // Move left pointer to the next alphanumeric character
            while (left < right && !Character.isLetterOrDigit(s.charAt(left))) {
                left++;
            }

            // Move right pointer to the previous alphanumeric character
            while (left < right && !Character.isLetterOrDigit(s.charAt(right))) {
                right--;
            }

            // Compare characters in a case-insensitive manner
            if (Character.toLowerCase(s.charAt(left)) != Character.toLowerCase(s.charAt(right))) {
                return false;
            }

            // Move pointers closer
            left++;
            right--;
        }

        // If no mismatches, return true
        return true;
    }
}

File: leetcode150-ts/Two Pointer/125_Valid_Palindrome.ts
/*
LeetCode 125. Valid Palindrome
Link: https://leetcode.com/problems/valid-palindrome/

A phrase is a palindrome if, after converting all uppercase letters into lowercase letters 
and removing all non-alphanumeric characters, it reads the same forward and backward.
*/

/**
 * Checks if a string is a valid palindrome after removing non-alphanumeric characters.
 * @param s - The input string to check
 * @returns boolean - True if the string is a palindrome, false otherwise
 */
function isPalindrome(s: string): boolean {
    // Step 1: Normalize the string
    // - Remove all non-alphanumeric characters using regex
    // - Convert to lowercase for case-insensitive comparison
    s = s.replace(/[^a-z0-9]/gi, '').toLowerCase();

    // Initialize two pointers:
    // - 'left' starts at the beginning of the string
    // - 'right' starts at the end of the string
    let left = 0;
    let right = s.length - 1;

    // Step 2: Check characters from both ends moving towards the center
    while (left < right) {
        // Compare characters at left and right pointers
        // - Increment left pointer if characters match
        // - Decrement right pointer if characters match
        // - Return false immediately if mismatch found
        if (s[left++] !== s[right--]) {
            return false;
        }
    }

    // If all character pairs matched, return true (valid palindrome)
    return true;
};

/*
================================================================================
THEORY & APPROACH EXPLANATION
================================================================================

1. PROBLEM UNDERSTANDING:
   - A palindrome reads the same forwards and backwards when considering only alphanumeric chars
   - We must ignore case and all non-alphanumeric characters
   - Examples:
     - "A man, a plan, a canal: Panama" → true ("amanaplanacanalpanama")
     - "race a car" → false ("raceacar")

2. KEY INSIGHTS:
   - The solution requires two main operations:
     a. String normalization (remove non-alphanumeric, lowercase)
     b. Palindrome verification

3. SOLUTION APPROACH:
   a. Normalization Phase:
      - Use regex to remove non-alphanumeric characters
      - Convert to lowercase for case-insensitive comparison
   b. Verification Phase:
      - Use two-pointer technique:
        - One pointer starts at beginning
        - One pointer starts at end
      - Compare characters while moving pointers towards center
      - If any mismatch found → not palindrome
      - If pointers meet/cross → valid palindrome

4. COMPLEXITY ANALYSIS:
   - Time Complexity: O(n)
     - replace(): O(n) for string scan
     - toLowerCase(): O(n)
     - Two-pointer check: O(n/2) → O(n)
   - Space Complexity: O(1)
     - Only uses pointers and creates one new string
     - No additional data structures used

5. EDGE CASES HANDLED:
   - Empty string → returns true (considered palindrome)
   - String with no alphanumeric chars → empty string → true
   - String with all same characters → returns true
   - Odd-length palindromes → middle character doesn't need match
*/

/*
================================================================================
TEST CASES
================================================================================
console.log(isPalindrome("A man, a plan, a canal: Panama")); // true
console.log(isPalindrome("race a car"));                    // false
console.log(isPalindrome(" "));                             // true
console.log(isPalindrome("No 'x' in Nixon"));               // true
console.log(isPalindrome("0P"));                            // false
*/